server/src/sim.rs

```diff
@@ -193,9 +197,96 @@ fn dump_entity(world: &hecs::World, entity: Entity) -> Vec<Component> {
     components
 }
 
+/// Ensure all nodes within `distance` of `start` exist
+pub fn ensure_nearby(graph: &mut Graph<Empty>, start: &Position, distance: f64) {
+    let mut pending = Vec::<(NodeId, na::Matrix4<f64>)>::new();
+    let mut visited = FxHashSet::<NodeId>::default();
+
+    pending.push((start.node, na::Matrix4::identity()));
+    visited.insert(start.node);
+    let start_p = start.local.map(|x| x as f64) * math::origin();
+
+    while let Some((node, current_transform)) = pending.pop() {
+        for side in Side::iter() {
+            let neighbor = graph.ensure_neighbor(node, side);
+            if visited.contains(&neighbor) {
+                continue;
+            }
+            visited.insert(neighbor);
+            let neighbor_transform = current_transform * side.reflection();
+            let neighbor_p = neighbor_transform * math::origin();
+            if math::distance(&start_p, &neighbor_p) > distance {
+                continue;
+            }
+            pending.push((neighbor, neighbor_transform));
+        }
+    }
+}
+
 struct Character {
     name: String,
     orientation: na::UnitQuaternion<f32>,
     direction: na::Unit<na::Vector3<f32>>,
     speed: f32,
 }
+
+#[cfg(test)]
+mod tests {
+
+    use super::*;
+
+    #[test]
+    fn rebuild_from_tree() {
```

Comment:
On review, I think these tests still belong in the `common` crate, since they're more about testing graph/cursor logic than the traversal. Maybe move `ensure_nearby` into a new module in `common` and restore these two tests to their original location?

Reply:
I've done as you requested. ensure_nearby is now found in traversal.rs in common, and the two tests I moved are back in their original locations in graph and cursor.

I'm not too clear on what the vision is for this change. If ensure_nearby is just going back to a module in common, why have it in a new module instead of just putting it in graph again? Should nearby_nodes also live in traversals?

Reply:
The general idea is to limit the visibility of things as narrowly as possible to the areas that need them. This applies both to the traversals themselves (so it's more obvious what they're for and more natural to make them even more application-specific in the future) and to the internals of `Graph`, which we're protecting from the traversal code that doesn't actually need them by making them no longer methods.